_posts/2025-03-20-phased-strimzi-upgrade-example.md

```diff
@@ -0,0 +1,637 @@
+
+Once we have applied this label to all the Kafka clusters in our fleet, we should see Strimzi 0.43 happily reconciling the Kafka CRs and the other Strimzi versions sitting idle.
+
+```mermaid
+flowchart TD
+	S_043(Strimzi 0.43) --> K_A(Kafka A)
+    S_043 --> K_B(Kafka B)
+    S_043 --> K_C(Kafka C)
+	S_044(Strimzi 0.44)
+	S_045(Strimzi 0.45)
+```
+
+We now want to upgrade to a newer Strimzi version. 
+The Kafka C cluster is our most risk tolerant cluster, it hosts topics used by low tier data pipelines which can tolerate extended downtime, so it will go first to ensure there are no issues.
+We first check that the Kafka C cluster is error free and in the `Ready` state and then apply the Strimzi custom resource selector label for the 0.44.0 Strimzi operator:
+
+```shell
+kubectl -n kafka label --overwrite kafka kafka-c strimzi-resource-selector=strimzi-0-44-0
+```
+
+The 0.44.0 Operator instance will see the new cluster and update the image reference.
+It will then roll the cluster, one broker at a time, on to the new image.
+
+```mermaid
+flowchart TD
+    S_043(Strimzi 0.43) --> K_A(Kafka A)
+    S_043 --> K_B(Kafka B)
+	S_044(Strimzi 0.44) --> K_C(Kafka C)
+	S_045(Strimzi 0.45)
+```
+
+After a burn-in period, where we are happy with the stability of the cluster, our client side canary jobs are stable and no client teams are reporting issues, we begin to move the rest of the fleet.
+
+```mermaid
+flowchart TD
+	S_043(Strimzi 0.43)
+    S_044(Strimzi 0.44) --> K_A(Kafka A)
+    S_044 --> K_B(Kafka B)
+	S_044 --> K_C(Kafka C)
+	S_045(Strimzi 0.45)
+	style K_A stroke:#f66,stroke-width:2px,stroke-dasharray: 5 5
+```
+
+All is well until we move our highest tier cluster, A. 
+The move to Strimzi 0.44 revealed a configuration issue.
+We think we know what the issue is and how to fix it, but as this is a critical cluster we need to roll back and make sure all the dependant services are running correctly. 
+In this setup the roll back is a simple label change:
+
+```shell
+kubectl -n kafka label --overwrite kafka kafka-a strimzi-resource-selector=strimzi-0-43-0
+```
+
+```mermaid
+flowchart TD
+    S_043(Strimzi 0.43) --> K_A(Kafka A)
+    S_044(Strimzi 0.44) --> K_B(Kafka B)
+	S_044 --> K_C(Kafka C)
+	S_045(Strimzi 0.45)
+```
+
+We fix the configuration issue and move Kafka cluster A over to the 0.44 operator.
+We then move cluster C onto the latest 0.45 operator, ready to upgrade the rest of the fleet.
+
+```mermaid
+flowchart TD
+    S_043(Strimzi 0.43)
+	style S_043 stroke:#f66,stroke-width:2px
+    S_044(Strimzi 0.44) --> K_A(Kafka A)
+	S_044 --> K_B(Kafka B)
+	S_045(Strimzi 0.45) --> K_C(Kafka C)
+```
+
+At this point the Strimzi 0.43 operator is not managing any Kafka clusters and all the clusters being managed by the 0.44 operator have been stable for sufficient period that we are confident we don't need to roll back. 
+We can now remove the Strimzi 0.43 operator deployment and its associated RBAC resources.
+
+### Things to watch out for
```

Comment:
You should add here also a warning about not removing the CRDs or ClusterRoles by mistake when deleting the old version. Especially with the CRDs, it would wipe all operands. This is the main issue of this approach.

Reply:
Yeah. Maybe it should also be mentioned at the beginning when dealing with pros and cons.

Reply:
It is definatly something to be mindful of so I will add it here. I'm not sure about adding at the beginning as it is a risk with a normal Strimzi deployment as well.

Reply:
I added this caveat to the summary section at the top of the post.